Add substitutions for pm25 correction formulas

MallocArray opened the pull request.

ypid reviewed (approved):
I have not reviewed the full C code but I still gave it a bit of thought. I like using `substitutions` for this. This keeps it very flexible. So it looks like a good way forward. I currently don’t have the time to test this but I will give it a shot at some point (months) because I think one of my sensors is also effected.

Mallonbacka commented:
I tried testing this manually, since I had the same algorithm configured a different way. I've only tested this from the point-of-view of someone with a sensor that needs adjustment, so I can't comment on how this works for others. I switched the relevant package files to use the equivalent from this branch and re-built and installed the firmware.

It all seems to work as described, thanks!

The only thing that was different was that you've used `accuracy_decimals: 0` for the new PM2.5 value. Maybe this makes sense, since if it were really measuring it, it would get a whole number, but, for whatever reason, I'd previously had `accuracy_decimals: 2` in my implementation. That means that after switching over, I get a history graph that looks like this:

<img width="1408" alt="Line graph showing the first few hours jumping erratically between lots of values between 3.5 and 5.5, then, after 17.00, moving less frequently between integer values 3, 4, 5 and 6" src="https://github.com/user-attachments/assets/62456dba-6daa-48df-bf80-330a9f4a5ee2" />

This isn't a problem for me, as I say, it better simulates having a working sensor, but a little insight into the trend is now missing. I am happy to get used to it if it's the way to go, though.

MallocArray commented:
The default setting from ESPHome for all of the sensors looks to be `accuracy_decimals: 0` even if not explicitly defined.  So I'm of the opinion to keep it the same for the calculated value, but looking at what my stock AirGradient is sending up to the server it looks like it is to 2 decimal points, so if wanting to match that, I could see the case.

With ESPHome, my understanding is that a new value is only written to the database when the value changes.  By keeping 0 decimal points, it would only change if it moves from 5 to 4, but if set to 2 decimal places, it will write for pretty much every update assuming it changes by that small degree.

Not sure how big of a deal that is for other users though.

So I could be talked into either way.

I still want to update the API to send the `pm02Compensated` value to the AG Dashboard and that is what I'm waiting for on this PR.

MallocArray commented:
Found the issue where it was still applying the formula to the 0.3 count values when they were less than 31, but my intention was that this would only be used if you specified a scaling_factor or intercept.

Added that as a conditional and now monitoring to make sure it behaves as expected.
Also added the corrected and raw PM2.5 values to the API payload to AirGradient
